### PR TITLE
Add alternate name for Plugin

### PR DESCRIPTION
**Pull request by dinujoh**
### Description
Add alternate name for Plugin
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by graytaylor0 — DISMISSED** [review]

File: data-prepper-plugin-framework/src/main/java/org/opensearch/dataprepper/plugin/ClasspathPluginProvider.java [hunk 1]
```diff
@@ -96,4 +98,18 @@ private void addOptionalDeprecatedPluginName(
             supportTypeToPluginTypeMap.put(supportedType, concretePluginClass);
         }
     }
+
+    private void addOptionalAlternatePluginName(
```
Comment (dlvenable) [inline data-prepper-plugin-framework/src/main/java/org/opensearch/dataprepper/plugin/ClasspathPluginProvider.java:102]: Let's combine this with `addOptionalDeprecatedPluginName` as it is pretty much the same logic.

```
    private void addPossiblePluginName(
            final Map<String, Map<Class<?>, Class<?>>> pluginsMap,
            final Class<?> concretePluginClass,
            final Function<DataPrepperPlugin, String> possiblePluginNameFunction,
            final Predicate<String> possiblePluginNamePredicate) {
        final DataPrepperPlugin dataPrepperPluginAnnotation = concretePluginClass.getAnnotation(DataPrepperPlugin.class);
        final String possiblePluginName = possiblePluginNameFunction.apply(dataPrepperPluginAnnotation);
        final Class<?> supportedType = dataPrepperPluginAnnotation.pluginType();

        if (possiblePluginNamePredicate.test(possiblePluginName)) {
            final Map<Class<?>, Class<?>> supportTypeToPluginTypeMap =
                    pluginsMap.computeIfAbsent(possiblePluginName, k -> new HashMap<>());
            supportTypeToPluginTypeMap.put(supportedType, concretePluginClass);
        }
    }

```

Also, I think we can use the same logic for the main name. But, please check that is correct first!

You can use with something like this:

```
addPossiblePluginName(pluginsMap, concretePluginClass, DataPrepperPlugin::name, deprecatedPluginName -> true);
addPossiblePluginName(pluginsMap, concretePluginClass, DataPrepperPlugin::deprecatedName, deprecatedPluginName -> !deprecatedPluginName.equals(DEFAULT_DEPRECATED_NAME));
addPossiblePluginName(pluginsMap, concretePluginClass, DataPrepperPlugin::alternateName, alternatePluginName -> !alternatePluginName.equals(DEFAULT_ALTERNATE_NAME));
```

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/annotations/DataPrepperPlugin.java [hunk 2]
```diff
@@ -46,6 +48,12 @@
      */
     String deprecatedName() default DEFAULT_DEPRECATED_NAME;
 
+    /**
+     *
+     * @return Alternate name of the plugin which should be unique for the type
+     */
+    String alternateName() default DEFAULT_ALTERNATE_NAME;
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/model/annotations/DataPrepperPlugin.java:55]: Let's make this an array. I could see this growing over time.

```
String[] alternateNames()
```

**Review by dlvenable — CHANGES_REQUESTED** [review]

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/annotations/DataPrepperPlugin.java (hunk 2 above)
Comment (dinujoh) [reply data-prepper-api/src/main/java/org/opensearch/dataprepper/model/annotations/DataPrepperPlugin.java]: Done

File: data-prepper-plugin-framework/src/main/java/org/opensearch/dataprepper/plugin/ClasspathPluginProvider.java (hunk 1 above)
Comment (dinujoh) [reply data-prepper-plugin-framework/src/main/java/org/opensearch/dataprepper/plugin/ClasspathPluginProvider.java]: Done

**Review by graytaylor0 — APPROVED** [review]

**Comment by graytaylor0** [comment]
Looks like the build is failing from a test in `data-prepper-api`

**Comment by dinujoh** [comment]
> Looks like the build is failing from a test in `data-prepper-api`



> Looks like the build is failing from a test in `data-prepper-api`

It might be flaky. Other java version build is succeeding. I will restart the test.

**Review by dlvenable — APPROVED** [review]

**Pull request merged**